File: Solutions/1642. Furthest Building You Can Reach/1642. Furthest Building You Can Reach.java
class Solution {
    public int furthestBuilding(int[] h, int bricks, int ladders) {
        PriorityQueue<Integer> pq=new PriorityQueue<>();  //min heap
        for(int i=0;i<h.length-1;++i){
            int dis=h[i+1]-h[i];
            if(dis>0)
                pq.offer(dis);
            if(pq.size()>ladders)
                bricks-=pq.poll();
            if(bricks<0)
                return i;
        }
        return h.length-1;
    }
}

File: Solutions/1642. Furthest Building You Can Reach/1642. Furthest Building You Can Reach.py
class Solution:
    def furthestBuilding(self, h: List[int], bricks: int, ladders: int) -> int:
        pq=[]
        for i in range(len(h)-1):
            dis=h[i+1]-h[i]
            if dis>0:
                heapq.heappush(pq,dis)
            if len(pq)>ladders:
                bricks-=heapq.heappop(pq)
            if bricks<0:
                return i
        return len(h)-1